### PR TITLE
BDOG-1690 Fix replaceIndexes for index rename

### PR DESCRIPTION
**Pull request by colin-lamed**

### CODE REVIEW

**Review by Shnick — DISMISSED** [review]

File: hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/MongoUtils.scala [hunk 1]
```diff
@@ -27,40 +27,49 @@ trait MongoUtils {
...
+                              Future.traverse(orphanedIndices) { indexName =>
+                                logger.warn(s"Index '$indexName' is not longer defined for ${collection.namespace}, removing")
+                                collection.dropIndex(indexName).toFuture()
+                                  .recoverWith {
+                                    // could be caused by race conditions between server instances
+                                    case IndexNotFound(e) => Future.unit
+                                  }
+                              }
+                            else
+                              Future.successful(logger.warn(s"The following indices exist in mongo for ${collection.namespace}, but are (no longer) provided to ensureIndexes: ${orphanedIndices.mkString(", ")}"))
```
Comment (chotai) [inline hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/MongoUtils.scala]: Although indices and indexes are both correct as the plural for index, as mongo has already fixed on indexes - and we have used indexes too (like in ensureIndexes), I feel we should just use indexes through out. I would replace all occurrences of indices with indexes.

Comment (chotai) [inline hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/MongoUtils.scala]: We should log the keys that make up the index as well imo - will be useful to detect where a index has been renamed and would be a good addition for the mongo 4.2 index change

Should we log the keys and indexes passed to ensureIndexes as well?

**Review by chotai — CHANGES_REQUESTED** [review]
If we have multiple instances of an app running, could we potentially run into issues where try to drop and recreate the orphaned indexes multiple times? Should we use a transaction here?

**Comment by colin-lamed** [comment]
It's not possible to edit indexes within a transaction. I can't really think of a way to avoid this (without mongo lock).
I think it would be very rare (race condition in the scenario that there's a deployment with an index update) and this approach is already being used (also from simple-reactivemongo) and I'm not aware of having seen this being an issue so far.

File: hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/MongoUtils.scala [hunk 2]
```diff
@@ -33,22 +33,31 @@ trait MongoUtils {
...
-                                    case IndexNotFound(e) => Future.unit
-                                  }
-                              }
-                            else
-                              Future.successful(logger.warn(s"The following indices exist in mongo for ${collection.namespace}, but are (no longer) provided to ensureIndexes: ${orphanedIndices.mkString(", ")}"))
-                          else
+      existingIndexes  <- collection.listIndexes().toFuture()
+      existingIndexMap =  existingIndexes.map(d => d("name").asString.getValue -> d).toMap
+      requestedIndexes =  indexes.map(_.getOptions.getName)
+      orphanedIndexes  =  existingIndexes
```
Comment (chotai) [inline hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/MongoUtils.scala]: could we use `existingIndexMap.keys()` here instead?

File: hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/MongoUtils.scala [hunk 3]
```diff
@@ -33,22 +33,31 @@ trait MongoUtils {
...
+                                  // could be caused by race conditions between server instances
+                                  case IndexNotFound(e) => Future.unit
+                                }
+                             }
+                          else if (orphanedIndexes.nonEmpty) {
+                            val str =
+                              orphanedIndexes
+                                .map(orphanName => s"index: '$orphanName' key: '${existingIndexMap(orphanName)("key")}'")
+                                .mkString(", ")
+                            Future.successful(logger.warn(s"The following indexes exist in mongo for ${collection.namespace}, but are (no longer) provided to ensure: $str"))
```
Comment (chotai) [inline hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/MongoUtils.scala]: should `ensure` be `ensureIndexes`?

File: hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/MongoUtils.scala [hunk 4]
```diff
@@ -27,40 +27,58 @@ trait MongoUtils {
...
+      existingIndexes  <- collection.listIndexes().toFuture()
+      existingIndexMap =  existingIndexes.map(d => d("name").asString.getValue -> d).toMap
+      requestedIndexes =  indexes.map(_.getOptions.getName)
+      orphanedIndexes  =  existingIndexes
+                            .map(_("name").asString.getValue)
+                            .toSet
+                            .diff(requestedIndexes.toSet + "_id_")
+      _                <- if (orphanedIndexes.nonEmpty && replaceIndexes)
+                            Future.traverse(orphanedIndexes) { orphanName =>
+                              logger.warn(s"Index '$orphanName' key: '${existingIndexMap(orphanName)("key")}' is no longer defined for ${collection.namespace}, removing")
```
Comment (chotai) [inline hmrc-mongo-common/src/main/scala/uk/gov/hmrc/mongo/MongoUtils.scala]: Wonder if `removing` should be made more explicit. eg `dropping index..`

**Review by chotai — DISMISSED** [review]

**Review by chotai — APPROVED** [review]

**Pull request merged**